### PR TITLE
atlasmigration: store `migrations.tar.gz` in k8s' secrets

### PR DESCRIPTION
**Pull request by giautm**

### CODE REVIEW

**Review by rotemtam — COMMENTED** [review]
Shouldn't we delete the previous version of the secret?

Also should secrets be deleted when the migration resource is deleted ?

**Comment by giautm** [comment]
> Also should secrets be deleted when the migration resource is deleted ?

Absolute yes.

> Shouldn't we delete the previous version of the secret?

I'm not sure when we need to delete the old version, because if we keep only single version, we only need to delete it when the migration resource was deleted.

If we store more than one version, we may delete the old one when a new deployment was success. But, what if the latest deployment is a `migrate down` - (with Aborted status?!) ? Or chain of migrate downs?

**Comment by rotemtam** [comment]
Chain of down migrations is an interesting use case. 
If you store that migration dir at the version, after the `migrate down` is done, I think that  it will be OK. 

E.g:
User deploys v3 - dir(1,2,3) is stored
The user asks to go down to v2 
Down migration applied - dir(1,2) is stored
The user asks to go down to v1
Down migration applied dir(1) is stored. 

Re: failed down migration, or aborted down migration, you never replace the stored state, since the state is still the version before you rolled back. 

Sounds OK ?

File: controllers/atlasmigration_controller.go [hunk 1]
```diff
@@ -173,12 +184,67 @@ func (r *AtlasMigrationReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+			return result(err)
+		}
+	}
 	status.ObservedHash = hash
 	res.SetReady(*status)
 	r.recorder.Eventf(res, corev1.EventTypeNormal, "Applied", "Version %s applied", status.LastAppliedVersion)
 	return ctrl.Result{}, nil
 }
 
+func (r *AtlasMigrationReconciler) readMigrations(ctx context.Context, res *dbv1alpha1.AtlasMigration) (migrate.Dir, error) {
```
Comment (rotemtam) [inline controllers/atlasmigration_controller.go:202]: migrations is a pretty overloaded term in this codebase. 
maybe `dirState`, e.g `readDirState`, `storeDirState` etc.?

**Comment by giautm** [comment]
>  failed down migration, or aborted down migration, you never replace the stored state, since the state is still the version before you rolled back.

This need to be check/update to expected in the PR https://github.com/ariga/atlas-operator/pull/167; for now, we just save the dir-state after successful reconcile, whatever flow it did.

File: controllers/atlasmigration_controller.go (hunk 1 above)
Comment (giautm) [reply controllers/atlasmigration_controller.go]: Updated;

File: controllers/atlasmigration_controller.go [hunk 2]
```diff
@@ -173,12 +184,67 @@ func (r *AtlasMigrationReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+		},
+	}
+	if err := r.Get(ctx, client.ObjectKeyFromObject(secret), secret); err != nil {
+		return nil, err
+	}
+	return extractDirFromSecret(secret)
+}
+
+func (r *AtlasMigrationReconciler) storeDirState(ctx context.Context, res *dbv1alpha1.AtlasMigration, dir migrate.Dir) error {
+	var labels = map[string]string{}
```
Comment (datdao) [inline controllers/atlasmigration_controller.go:216]:
```suggestion
	var labels = make(map[string]string, len(res.labels))
```

**Review by a8m — APPROVED** [review]

**Pull request merged**